derive/src/lib.rs

```diff
@@ -97,33 +119,44 @@ fn generate_type(input: TokenStream2) -> Result<TokenStream2> {
     let generic_type_ids = ast.generics.type_params().map(|ty| {
         let ty_ident = &ty.ident;
         quote! {
-            :: #scale_info ::meta_type::<#ty_ident>()
+            #scale_info::meta_type::<#ty_ident>()
```

Comment:
as before for all non-`scale-info` crates this should still prepend `::` to signal root crate path.

Reply:
It doesn't compile with the `::` and I don't know why tbh, but I think that when deriving for `scale-info` itself, and we have `extern crate self as _scale_info;`, then `::_scale_info::path::to::something` doesn't work. :/

Reply:
The `extern crate self as _scale_info;` is not required since Rust edition 2018 and I don't think we should support Rust edition 2015 tbh because it actually resolves many problems we are facing otherwise such as this one.

Reply:
So either this crate should produce: `crate::` paths or whatever crate alias a dependency uses for `scale-info` prepended with `::`.

Reply:
To be more concrete: While the `proc-macro-crate` crate is actually useful in these cases it makes use of file I/O while expanding a proc. macro. There are discussions about encapsulating proc. macros in stricter environments that would break proc. macros such as these. Generally tooling such as `rust-analyzer` has good reasons why proc. macros that are "impure" should not be relied on.

Reply:
> To be honest I am not sure how counterintuitive it is to use the derive macros for the crate that exposes them.

How do you mean "counterintuitive"? I take it you mean it's forcing things a bit?

I think it's a good litmus test for the crate, especially since the type zoo in `scale-info` is fairly advanced and gives us a "free test suite" of sorts. The amount of manual implementations we'd have to maintain is pretty large so if we can derive most of them that's a win imo.

I am concerned by the increased complexity this brings. I will see what I can do to make things a wee bit less messy.

Reply:
> I am concerned by the increased complexity this brings. I will see what I can do to make things a wee bit less messy.

I think [this commit](https://github.com/paritytech/scale-info/pull/72/commits/b1d26c27db55b638dbfbfaa6f4ef943fff04829a) is makes things much cleaner.

Reply:
Proc. macros bring overhead with them. So some people prefer using crates but not their provided proc. macros which is why we find `derive` crate feature on some crates to opt-out. It is just nice to be able to separate concerns between the crate that provides definitions and the crate that provides the proc. macro implementations. As soon as the crate providing definitions makes use of the proc. macro you no longer have the separation and a direct dependency instead. I personally am not against this but overhead can be serious in some cases.
The biggest problem is that derive macros that can be used today in their parent crate require this type of "hack" that is encapsulated by the `proc-macro-crate` crate currently. Using this "hack" in any proc. macro turns the proc. macro impure. For example it can no longer enjoy acceleration by the techniques behind the `watt` crate which some people believe to be the future compilation model for all proc. macros since it is super fast and more secure than today's proc. macros.

Reply:
1. The use of `proc-macro-crate` predates this PR so the technical debt is already there;
2. This PR does not introduce a mandatory dependence on `scale-info-derive`: it's opt-in through a feature (as it should be, couldn't agree with you more);

I think we should have a conversation about the use of `proc-macro-crate`; you make a really good point about it making the crate "impure" and we should hash out the trade-offs involved. But I'm not sure this PR is the right place for that?